shared/src/main/java/org/odk/collect/shared/locks/ThreadSafeBooleanChangeLock.kt

```diff
@@ -0,0 +1,44 @@
+package org.odk.collect.shared.locks
+
+import java.util.function.Function
+
+class ThreadSafeBooleanChangeLock : ChangeLock {
+    private var locked = false
+
+    override fun <T> withLock(function: Function<Boolean, T>): T {
+        val acquired = tryLock()
+
+        return try {
+            function.apply(acquired)
+        } finally {
+            if (acquired) {
+                unlock()
+            }
+        }
+    }
+
+    override fun tryLock(): Boolean {
+        return synchronized(this) {
+            if (locked) {
+                false
+            } else {
+                locked = true
+                true
+            }
+        }
+    }
+
+    override fun lock() {
+        if (locked) {
```

Comment:
This should be wrapped in a `synchronized` to make sure that the lock can't be acquired between the `if` check and the assignment.

Reply:
Done.